### ISSUE TITLE
Web based configuration script

### ISSUE THREAD

**Issue by jasonmunro** [cypht-org/cypht#17]
Currently we have a set of CLI scripts to create the Cypht configuration and add/remove/modify user accounts (when using the DB authentication method). It would be great to have a web based version of these to make installation easier, especially for shared hosting environments.

**Comment by dumblob** [comment]
In my opinion this is quite a minor issue. In fact, this is a thing of a system-specific configuration tools or principles. It might be some packaging system (rpm, deb, pkgbuild, ebuild, etc.), it might be some configuration system (ansible, chef, puppet, etc.), it might be some specific GUI application, it might be just a text editor (nvim, emacs, nano, etc.). The only solution which from my experience works is well-defined configuration file semantics with well-known structure (e.g. JSON) at a well-known location (e.g. `/etc/`).

**Comment by jasonmunro** [comment]
I agree, but this is a big blocker to adoption, and if people can't try out the software it's hard to build momentum behind it. We include CLI helper scripts for a few things already, it's not too much of a stretch to "web-ify" them. My main concern with both the configuration arrangement we already have and a web based setup script is security.

**Comment by dumblob** [comment]
Right, what I meant was *"begin with a thoroughly defined configuration file (in terms of semantics and syntax)"* as the core feature and *"end with a simple web GUI operating upon this well-defined configuration file"*.

And it's true, that unification is the way to go. Not a set of different user-space scripts with varying interfaces.

Regarding security, do you mean the need to introduce authentication and at the same time providing bug-free web app being safe when reading/writing the configuration file(s)?

**Comment by jasonmunro** [comment]
thanks for the clarification. That is very much the approach I'm taking. The "defined" file is the hm3.ini file, the output is the rc file (serialized array). The web interface is just a way to manage both without a text editor and cli commands, but it ultimately will use the same code path.

WRT security, the issue is authentication. CLI commands have an implicit authentication associated with them (even that is not ideal), but providing a web interface to "bootstrap" an installation is worrisome, if for no other reason than it gives people an easy way to shoot themselves in the foot :)

**Comment by Kroesss** [comment]
Is there any news on this ticket yet? It would be great if it could be used on a shared hosting-environment.

**Comment by jasonmunro** [comment]
@Kroesss, Sorry, this is still pretty low on my priority list right now.

**Comment by jasonmunro** [comment]
As of today I have disabled the half ass web based configuration builder we had in the scripts directory. I did for a number of reasons as described in the commit:

1. It does not work
2. While only available in debug mode, it's not secure
3. Even if I get it working, it's not going to be easy to use and hard to maintain.
4. I'm seriously considering trashing the idea completely

As much as I understand why people want a web based installer, I just don't think it makes sense with the Cypht install process.

**Comment by marclaporte** [comment]
For folks that need this today: Cypht is bundled in Tiki, and Tiki is shared hosting friendly since 2002! As a bonus, you get CardDAV and CalDAV support (Beta).
https://info.tiki.org/Benefits
https://dev.tiki.org/Cypht-integration

**Comment by Yamakasi** [comment]
> 
> 
> For folks that need this today: Cypht is bundled in Tiki, and Tiki is shared hosting friendly since 2002! As a bonus, you get CardDAV and CalDAV support (Beta).
> https://info.tiki.org/Benefits
> https://dev.tiki.org/Cypht-integration

What kind of advertisement is this? Please stop that.

Thanks you!

**Comment by marclaporte** [comment]
Here is why this is relevant information:

The topic of this issue is to be able to install Cypht via a web installer like most PHP web apps. So the typical use case of a web control panel like Virtualmin, ISPconfig or the well-known proprietary alternatives.
# Create DB (via the web control panel)
# Get the source (ex.: via a .zip)
# Unzip the source in a web accessible folder and point a browser to it.
# You will then get a web installer. Answer the questions (ex.: the previously created database info)

The fact that this is missing is "a big blocker to adoption" as stated by Jason.

AFAIK, there is only one way today to get Cypht with a web-based installer, and it's via Tiki. Tiki has tons of features but they are all optional. So you can activate only Cypht. So basically, Cypht can use Tiki's installer. Cypht and Tiki share the same license and programming language, so it's even easy to move/copy code as appropriate.

We are "upstream first" as explained here:
https://www.redhat.com/en/blog/upstream-first-turning-openstack-nfv-platform
Every webmail enhancement Tiki makes is first proposed to be in Cypht. And if it's out of scope, it will go in Tiki only. Here are some discussions about scope:
* https://github.com/jasonmunro/cypht/issues/382
* https://github.com/jasonmunro/cypht/issues/115

We should make a Cypht Webmail profile to make it even easier: https://profiles.tiki.org/

And my goal is to get other PHP web apps to adopt Cypht in the same way. The first opens the way. The more projects do, the more eyeballs and developers we have. Cypht + Packagist is for Cypht, for Tiki and for all future projects that will embed Cypht:
https://github.com/jasonmunro/cypht/issues/311

In a nutshell:
# Cypht is awesome
# We'd like Cypht to be easier to install
# One great way is for Cypht to have a web-based installer.
# From Jason's last comment, it seems unlikely to happen
# My suggestion is a real world workaround available now
# It does not stop any effort to create a Cypht-specific web installer
# All webmail enhancements are done directly in Cypht when appropriate. So Cypht and Tiki have a win-win relationship. There is no downside or tradeoff.

Tiki has over 1 million downloads and it is promoting Cypht, so it makes perfect sense for Cypht to promote Tiki as well.

Best regards,

**Comment by marclaporte** [comment]
I was thinking about this yesterday. Beyond a standard installer, we should have something to set up tests:
https://github.com/jasonmunro/cypht/issues/596

Perhaps this could be done with Composer?

**Comment by marclaporte** [comment]
A few years back, we did some research for https://dev.tiki.org/Generic-PHP-app-deployment-tools-which-are-written-in-PHP
In the end, we opted to continue work on our Tiki-specific tool: https://doc.tiki.org/Manager (which started in 2008 or so)

When the time comes to work on this, let's see if we can reuse an existing project. Let's make it super easy for web hosting providers and control panels to provide and maintain Cypht.

**Comment by marclaporte** [comment]
Here are examples of recipes: https://github.com/deployphp/deployer/tree/master/recipe

**Comment by marclaporte** [comment]
Related: https://github.com/YunoHost-Apps/cypht_ynh

**Comment by marclaporte** [comment]
We'll work on improving the process to avoid this:
"I'm really totally lost with this, my best best would have been [Cypht](https://cypht.org/), but after days of trying to install it, I'm about to give up because I just can't get it to work (stuck at the login screen and doesn't go further although everything is setup and configured properly)." Source:
https://github.com/roundcube/roundcubemail/issues/4972#issuecomment-640747464

**Comment by marclaporte** [comment]
I wonder: Should we invest our time in a web-based installer or improve our Docker package?

**Comment by marclaporte** [comment]
We are working on the docs here: https://github.com/cypht-org/cypht-website/pull/46

**Comment by marclaporte** [comment]
> mprove our Docker package

We have massively improved our Docker packaging: https://github.com/cypht-org/cypht-docker/issues/31

**Comment by IrAlfred** [comment]
Due to security concerns related to the web-based installation method for Cypht, which would require a thorough security analysis, and because this is a low-priority issue (as root access is needed regardless), we are closing this for now. The standard installation process for Cypht is already well-documented and simplified on the official website [https://cypht.org/install](https://cypht.org/install), which covers various contexts including manual installation, Docker, Tiki, and YunoHost.

**Comment by marclaporte** [comment]
> as root access is needed regardless

@IrAlfred Why is root access needed? Is there not a workaround?

**Comment by marclaporte** [comment]
A user wrote "What I'm getting from this is that, unfortunately, the idea of a lightweight, easy to deploy PHP webmail client is dead, with no like for like alternatives on the horizon?" Source: https://github.com/the-djmaze/snappymail/issues/1911#issuecomment-3960544535

What can we learn from Snappymail?
https://github.com/the-djmaze/snappymail/wiki/Installation-instructions

Looking at the options on https://wikisuite.org/Webmail-and-groupware-comparison, the other active, standalone (and thus that require an installer) and in PHP are Roundcube and Cypht.

Here is the doc for Roundcube, (the most popular PHP webmail):
https://github.com/roundcube/roundcubemail/blob/master/docs/INSTALL.md

Cypht is a great PHP webmail but it's not easy enough to install for traditional shared web hosting.
https://www.cypht.org/install/

I accept that this is low priority as we have better docs, along with Docker and Yunohost installs. But eventually, we should have an easier shared hosting option. So I am re-opening this issue.

**Comment by IrAlfred** [comment]
> > as root access is needed regardless
> 
> [@IrAlfred](https://github.com/IrAlfred) Why is root access needed? Is there not a workaround?

It's not strictly mandatory, but for a secure setup as recommended here https://github.com/cypht-org/cypht/issues/1861#issuecomment-3912621642, the Cypht source files should be placed outside the document root (e.g., in /usr/local/share/cypht/). In this specific case, doing so requires root access to create the necessary folder. So ultimately, it can depend on the location you choose.

**Comment by IrAlfred** [comment]
> What can we learn from Snappymail?

We will try to explore this (Snappymail and roundcube) to learn something from their process

**Comment by marclaporte** [comment]
> It's not strictly mandatory

Understood. We should lower barriers to join our community.

The default documentation should have a recipe for non-root (a typical shared hosting setup).

I suggest to spin up a lab server with https://wikisuite.org/Virtualmin as a reference implementation.

You will can then have something like:
- home/abc/cypht (outside web root)
- home/abc/public_html (web-root)


The target: I recommend to find a non-techie friend or family member and make sure the documentation is clear enough that they can install Cypht.

**Comment by marclaporte** [comment]
Suggested again: https://github.com/cypht-org/cypht/issues/1861#issuecomment-4933002657

**Comment by Shadow243** [comment]
I will push a fix for this
